#! /bin/bash

# sampleId= ### determined from the runScript
# bamFile=
# Initial BAM file from novoalign -> picard

unset realignerIntervalFile bamRealigned metricsFile bamMarkDup intervalFile grpFile postGrpFile plotFile bamRecal callConf emitConf nt allRawOut snpRawOut indelRawOut recalFile tranchesFile rscriptFile tsFilterLevel snpRecalFiltVcf indel_hardFileVcf allFiltVcf

realignerIntervalFile='020_refineAlignment/010_realignGATK/'$sampleId'_realigner.intervals'
# realignerIntervalFile contain the small suspicious intervals which are likely in need of realignment detected by gatk RealignerTargetCreator.
 
bamRealigned='020_refineAlignment/010_realignGATK/'$sampleId'_indelRealigned.bam'
# bam file from gatk IndelRealigner
 
metricsFile='020_refineAlignment/020_markDupPicard/'$sampleId'_markDup.metrics.txt'
 
bamMarkDup='020_refineAlignment/020_markDupPicard/'$sampleId'_MarkedDup.bam'
#bam file from picard markduplicates
 
intervalFile="/home/pubuduss/nexterarapidcapture_exome_targetedregions_v1.2.list"
#One or more genomic intervals over which to operate
 
grpFile='020_refineAlignment/030_BQRecalGATK/'$sampleId'_recal_data.grp'
#table with recalibration data
 
postGrpFile='020_refineAlignment/030_BQRecalGATK/'$sampleId'_post_recal_data.grp'
#table used to plot the improvements of base recalibration.
 
plotFile='020_refineAlignment/030_BQRecalGATK/'$sampleId'_bqsr_plots.pdf'
 
bamRecal='060_delivery/'$sampleId'_all.realigned.markDup.baseQreCali.bam'
# Final realigned markeddup recalibrated bam file
 
callConf=50
# -stand_call_conf value
 
emitConf=10
# --stand_emit_conf value
 
nt=3;
# how many threads can be used
 
allRawOut='040_variantCalling/gatk/'$sampleId'_all.raw.vcf'
#A raw, unfiltered, highly sensitive callset in VCF format
 
snpRawOut='040_variantCalling/gatk/'$sampleId'_snp.raw.vcf'
#Extracted SNPs from $allRawOut
 
indelRawOut='040_variantCalling/gatk/'$sampleId'_indel.raw.vcf'
#Extracted InDels from $allRawOut
 
recalFile='050_postVarCalProcess/gatk/010_qualityFiltration/'$sampleId'_snp.vqsr.output.recal'
#The output recal file of SNPs used by ApplyRecalibration
 
tranchesFile='050_postVarCalProcess/gatk/010_qualityFiltration/'$sampleId'_snp.vqsr.output.tranches'
#The output tranches file used by ApplyRecalibration (file describing where to cut the data)
 
rscriptFile='050_postVarCalProcess/gatk/010_qualityFiltration/'$sampleId'_snp.vqsr.output.plots.R'
#The output rscript file generated by the VQSR to aid in visualization of the input data and learned model
 
tsFilterLevel=99
#The truth sensitivity level at which to start filtering
 
snpRecalFiltVcf='050_postVarCalProcess/gatk/010_qualityFiltration/'$sampleId'_snp.recalibrated.filtered.vcf'
#A recalibrated VCF file in which each variant is annotated with its VQSLOD and filtered if the score is below the desired quality level.
 
indel_hardFileVcf='050_postVarCalProcess/gatk/010_qualityFiltration/'$sampleId'indel.hardFiltered.vcf'
#File from indel hard filtration
 
allFiltVcf='050_postVarCalProcess/gatk/010_qualityFiltration/'$sampleId'_all.filter.vcf'

aviFile='050_postVarCalProcess/020_annovarAnnotation/'$sampleId'avinput'
